infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -46,58 +56,137 @@
     this.clusterRoleName = clusterRoleName;
     this.clientFactory = clientFactory;
     this.defaultNamespaceName = defaultNamespaceName;
     this.allowUserDefinedNamespaces = allowUserDefinedNamespaces;
+    this.workspaceManager = workspaceManager;
+
+    // This will disappear once we support user selection of workspaces...
+    if (allowUserDefinedNamespaces) {
+      LOG.warn(
+          "'che.infra.kubernetes.namespace.allow_user_defined' is not supported yet. It currently has no"
+              + " effect.");
+    }
+
+    // right now allowUserDefinedNamespaces can't be true, but eventually we will implement it.
     if (isNullOrEmpty(defaultNamespaceName) && !allowUserDefinedNamespaces) {
       throw new ConfigurationException(
           "che.infra.kubernetes.namespace.default or "
               + "che.infra.kubernetes.namespace.allow_user_defined must be configured");
     }
   }
 
-  private boolean hasNoPlaceholders(String namespaceName) {
-    return NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().noneMatch(namespaceName::contains);
+  private boolean hasPlaceholders(String namespaceName) {
+    return namespaceName != null
+        && NAMESPACE_NAME_PLACEHOLDERS.keySet().stream().anyMatch(namespaceName::contains);
   }
 
   /**
-   * True if namespace is predefined for all workspaces. False if each workspace will be provided
-   * with a new namespace or provided for each user when using placeholders.
+   * True if namespace is potentially created for the workspace, false otherwise.
+   *
+   * <p>The logic is a little bit non-trivial and best expressed by just fully evaluting the truth
+   * table as below ({@code ...namespace} stands for the legacy namespace property, {@code
+   * ...namespace.default} stands for the namespace default property):
+   *
+   * <pre>{@code
+   * ...namespace    | ...namespace exists | ...namespace.default | creating?
+   * no-placeholders |       no            |       null           | error
+   * no-placeholders |       no            |   no-placeholders    | no
+   * no-placeholders |       no            |    placeholders      | yes
+   * no-placeholders |      yes            |       null           | no
+   * no-placeholders |      yes            |   no-placeholders    | no
+   * no-placeholders |      yes            |    placeholders      | no
+   *  placeholders   |       no            |        null          | error
+   *  placeholders   |       no            |   no-placeholders    | no
+   *  placeholders   |       no            |    placeholders      | yes
+   *  placeholders   |      yes            |        null          | yes
+   *  placeholders   |      yes            |   no-placeholders    | yes
+   *  placeholders   |      yes            |    placeholders      | yes
+   * }</pre>
    */
-  public boolean isPredefined() {
-    return isPredefined;
+  protected boolean isCreatingNamespace(String workspaceId) throws InfrastructureException {
```

Comment:
I propose to refactor this method in the following way
```java
  protected boolean isCreatingNamespace(String workspaceId) throws InfrastructureException {
    boolean legacyExists =
        checkNamespaceExists(
            resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId));

    // legacy namespace exists and should be used
    if (legacyExists) {
      // if it contains any placeholder - it indicates that Che created namespace by itself
      return isNullOrEmpty(legacyNamespaceName) || hasPlaceholders(legacyNamespaceName);
    }

    if (isNullOrEmpty(defaultNamespaceName)) {
      throw new InfrastructureException(
          "Cannot determine whether a new namespace and service account should be"
              + " created for workspace %s. There is no pre-existing workspace namespace to be found using the legacy"
              + " `che.infra.kubernetes.namespace` property yet the `che.infra.kubernetes.namespace.default` property"
              + " is undefined.");
    }

    return hasPlaceholders(defaultNamespaceName);
  }
```

Reply:
That is indeed much nicer to look at. Thanks!